contracts/vending/VendingMachine.sol

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "@thesis/solidity-contracts/contracts/token/IReceiveApproval.sol";
+import "../token/T.sol";
+
+/// @title T token vending machine
+/// @notice Contract implements a special update protocol to enable KEEP/NU
+///         token holders to wrap their tokens and obtain T tokens according
+///         to a fixed ratio. This will go on indefinitely and enable NU and
+///         KEEP token holders to join T network without needing to buy or
+///         sell any assets. Logistically, anyone holding NU or KEEP can wrap
+///         those assets in order to receive T. They can also unwrap T in
+///         order to go back to the underlying asset. There is a separate
+///         instance of this contract deployed for KEEP holders and a separate
+///         instance of this contract deployed for NU holders.
+contract VendingMachine is Ownable, IReceiveApproval {
+    using SafeERC20 for IERC20;
+    using SafeERC20 for T;
+
+    /// @notice Divisor for precision purposes, used to represent fractions.
+    uint256 public constant FLOATING_POINT_DIVISOR = 1e18;
+
+    /// @notice The token being wrapped to T (KEEP/NU).
+    IERC20 public immutable wrappedToken;
+
+    /// @notice T token contract.
+    T public immutable tToken;
+
+    /// @notice The ratio with which T token is converted based on the provided
+    ///         token being wrapped (KEEP/NU), expressed in 1e18 precision.
+    ///
+    ///         When wrapping:
+    ///           x [T] = amount [KEEP/NU] * ratio / FLOATING_POINT_DIVISOR
+    ///
+    ///         When unwrapping:
+    ///           x [KEEP/NU] = amount [T] * FLOATING_POINT_DIVISOR / ratio
+    uint256 public immutable ratio;
+
+    /// @notice The total balance of wrapped tokens for the given holder
+    ///         account. Only holders that have previously wrapped KEEP/NU to T
+    ///         can unwrap, up to the amount previously wrapped.
+    mapping(address => uint256) public wrappedBalance;
+
+    event Wrapped(
+        address indexed recipient,
+        uint256 wrappedTokenAmount,
+        uint256 tTokenAmount
+    );
+    event Unwrapped(
+        address indexed recipient,
+        uint256 tTokenAmount,
+        uint256 wrappedTokenAmount
+    );
```

Comment:
To make it perfect we can add docstrings for events too (and generate docs from contracts in some moment in a future)

Reply:
Opened #11